### PR TITLE
Fix: Date math with Interval keyword

### LINKED ISSUE #12076: Bug Report: Date math produces syntax error
### Overview of the Issue

Vitess 14 doesn't support some set of date math queries, for example: `select '2020-01-01' + interval month(DATE_SUB(FROM_UNIXTIME(1234), interval 1 month))-1 month` is a valid mysql query, but produces a syntax error when run against Vitess 14

Workaround is to use parens: `select '2020-01-01' + interval (month(DATE_SUB(FROM_UNIXTIME(1234), interval 1 month))-1) month` but original query shouldn't fail like this

### Reproduction Steps

Run `select '2020-01-01' + interval month(DATE_SUB(FROM_UNIXTIME(1234), interval 1 month))-1 month`

Another example: `select DATE_ADD(MIN(FROM_UNIXTIME(1673444922)),interval -DAYOFWEEK(MIN(FROM_UNIXTIME(1673444922)))+1 DAY)`

### Binary Version

```sh
Vitess 14.04
```

### Operating System and Environment details

```sh
Currently being reproduced with java unit tests against vtcombo
```

### Log Fragments

```sh
org.jdbi.v3.core.statement.UnableToExecuteStatementException: java.sql.SQLSyntaxErrorException: message: "syntax error at position 120"
code: INVALID_ARGUMENT
 [statement:"select /*+ MAX_EXECUTION_TIME(30000) */ '2020-01-01' + interval month(DATE_SUB(FROM_UNIXTIME(1234), interval 1 month))-1 month", arguments:{positional:{}, named:{}, finder:[]}]
```

```
org.jdbi.v3.core.statement.UnableToExecuteStatementException: java.sql.SQLSyntaxErrorException: message: "syntax error at position 133"
code: INVALID_ARGUMENT
 [statement:"select /*+ MAX_EXECUTION_TIME(30000) */ DATE_ADD(MIN(FROM_UNIXTIME(1673444922)),interval -DAYOFWEEK(MIN(FROM_UNIXTIME(1673444922)))+1 DAY)", arguments:{positional:{}, named:{}, finder:[]}]
```
```

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

This PR fixes the issue described in #12076.
The queries which are valid but didn't work with Vitess have been added as tests and the parser has been patched.

## Related Issue(s)

- Fixes #12076 

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by GuptaManan100** [comment]
I am unsure about the backport to any supported release. This syntax isn't really a regression, but it is still a bug fix.

**Comment by frouioui** [comment]
> This syntax isn't really a regression, but it is still a bug fix.

I think for this reason, let's not backport it @GuptaManan100.

**Review by frouioui — APPROVED** [review]

**Comment by olyazavr** [comment]
Would it be possible to backport this to v14? It has already appeared in a few places and broken people's queries, making it very difficult for us to upgrade to v14

File: go/test/endtoend/vtgate/queries/misc/misc_test.go [hunk 1]
```diff
@@ -161,3 +161,12 @@ func TestQueryTimeoutWithTables(t *testing.T) {
...
 }
+
+// TestIntervalWithMathFunctions tests that the Interval keyword can be used with math functions.
+func TestIntervalWithMathFunctions(t *testing.T) {
+	mcmp, closer := start(t)
+	defer closer()
+
+	mcmp.AssertMatches("select '2020-01-01' + interval month(DATE_SUB(FROM_UNIXTIME(1234), interval 1 month))-1 month", `[[CHAR("2020-12-01")]]`)
+	mcmp.AssertMatches("select DATE_ADD(MIN(FROM_UNIXTIME(1673444922)),interval -DAYOFWEEK(MIN(FROM_UNIXTIME(1673444922)))+1 DAY)", `[[DATETIME("2023-01-08 19:18:42")]]`)
+}
```
Comment (dbussink) [inline go/test/endtoend/vtgate/queries/misc/misc_test.go:172]: @GuptaManan100 Looks like this new test is failing? 

```
=== RUN   TestIntervalWithMathFunctions
[1648](https://github.com/vitessio/vitess/actions/runs/3899547338/jobs/6659836996#step:12:1649)
    misc_test.go:171: Query: select DATE_ADD(MIN(FROM_UNIXTIME(1673444922)),interval -DAYOFWEEK(MIN(FROM_UNIXTIME(1673444922)))+1 DAY) (-want +got):
[1649](https://github.com/vitessio/vitess/actions/runs/3899547338/jobs/6659836996#step:12:1650)
          strings.Join({
[1650](https://github.com/vitessio/vitess/actions/runs/3899547338/jobs/6659836996#step:12:1651)
          	`[[DATETIME("2023-01-08 1`,
[1651](https://github.com/vitessio/vitess/actions/runs/3899547338/jobs/6659836996#step:12:1652)
        - 	"9:1",
[1652](https://github.com/vitessio/vitess/actions/runs/3899547338/jobs/6659836996#step:12:1653)
        + 	"3:4",
[1653](https://github.com/vitessio/vitess/actions/runs/3899547338/jobs/6659836996#step:12:1654)
          	`8:42")]]`,
[1654](https://github.com/vitessio/vitess/actions/runs/3899547338/jobs/6659836996#step:12:1655)
          }, "")
[1655](https://github.com/vitessio/vitess/actions/runs/3899547338/jobs/6659836996#step:12:1656)
        
[1656](https://github.com/vitessio/vitess/actions/runs/3899547338/jobs/6659836996#step:12:1657)
        Got:[[DATETIME("2023-01-08 13:48:42")]]
[1657](https://github.com/vitessio/vitess/actions/runs/3899547338/jobs/6659836996#step:12:1658)
--- FAIL: TestIntervalWithMathFunctions (0.02s)
```

**Comment by olyazavr** [comment]
This works for our test cases (ported it over and ran vtcombo with the unit tests that were previously failing)

**Comment by GuptaManan100** [comment]
@dbussink I am looking into the failure. This is very weird though. The test seems to work for me locally!

**Comment by GuptaManan100** [comment]
Oh I just figured it out! Lol, the failure is because of different time zones. The UTC value is interpreted correctly, but the system time-zone is used to print the output. For me the timezone is IST which is +5:30, so it prints a time like `19:18:42`, but the time-zone of the github-runner is different, so it prints `13:48:42` which corresponds to +0:00

**Comment by dbussink** [comment]
> Oh I just figured it out! Lol, the failure is because of different time zones. The UTC value is interpreted correctly, but the system time-zone is used to print the output. For me the timezone is IST which is +5:30, so it prints a time like `19:18:42`, but the time-zone of the github-runner is different, so it prints `13:48:42` which corresponds to +0:00

Ah yeah, we probably have to assume UTC for things like this? And enforce that in the test?

**Comment by GuptaManan100** [comment]
I set the time-zone explicitly in the test now. Its a session setting which can be changed and this is something that Vitess allows to change too, so all good!

**Review by dbussink — APPROVED** [review]

**Comment by GuptaManan100** [comment]
The backports are - #12102, #12103 and #12106

**Pull request merged**